### PR TITLE
New deepcopy semantics

### PR DESCRIPTION
**Pull request by phschiele**
## Description
Resumes #2014. This PR can be used to work on the new semantics, and we can still decide which version to backport to 1.1, 1.2, 1.3, respectively. 

Implementation based on a suggestion by @michael-123123.

TODO: Documentation for new behaviour. 

## Type of change
- [x] New feature (backwards compatible)
- [ ] New feature (breaking API changes)
- [ ] Bug fix
- [ ] Other (Documentation, CI, ...)

## [Contribution checklist](https://www.cvxpy.org/contributing/index.html#contribution-checklist)
- [x] Add our license to new files.
- [x] Check that your code adheres to our coding style.
- [x] Write unittests.
- [x] Run the unittests and check that they’re passing.
- [x] Run the benchmarks to make sure your change doesn’t introduce a regression.

### CODE REVIEW

File: cvxpy/tests/test_copy.py [hunk 1]
```diff
@@ -0,0 +1,122 @@
...
+    x = cp.Variable()
+
+    a = Equality(x, 0)
+    b = copy.copy(a)
+    c = copy.deepcopy(a)
+
+    assert a.id == b.id
+    assert a.id != c.id
+
+    assert id(a) != id(b)
```
Comment (phschiele) [inline cvxpy/tests/test_copy.py:47]: Constraints have special handling in that after a copy `.id` matches but `id()` doesn't.

Comment (SteveDiamond) [reply cvxpy/tests/test_copy.py:47]: I think this is ok. Everything matches after deepcopy as expected.

**Comment by SteveDiamond** [comment]
This looks fine to me, once there's some documentation. The memo field in deepcopy reminds of how convex.jl would memoize during canonicalization, so if the same expression appears multiple times it will only be canonicalized once. That might actually be quite easy to add.

File: cvxpy/tests/test_copy.py (hunk 1 above)
Comment (SteveDiamond) [reply cvxpy/tests/test_copy.py:47]: It's ok IMO. I guess the idea here is a constraint is not a leaf, ``copy`` does need to rebuild it.

Comment (phschiele) [reply cvxpy/tests/test_copy.py:47]: Makes sense, so I guess my question would have been why `.id` remains the same. Not suggesting that it should be changed, though.

Comment (SteveDiamond) [reply cvxpy/tests/test_copy.py:47]: Constraint.id is used to propagate dual variables. So it's part of constraint.get_data, which is preserved by copy.

**Review by SteveDiamond — APPROVED** [review]
LGTM

File: cvxpy/tests/test_copy.py [hunk 2]
```diff
@@ -0,0 +1,122 @@
...
+    assert id(b) != id(c)
+
+
+def test_problem():
+    x = cp.Variable()
+    y = cp.Variable()
+    obj = cp.Minimize((x + y) ** 2)
+    constraints = [x + y == 1]
+    prob = cp.Problem(obj, constraints)
+    prob_copy = prob.copy()
```
Comment (PTNobel) [inline cvxpy/tests/test_copy.py:73]: Documentation mentions `copy.copy` but we're only testing `Problem.copy`. I realize `__copy__` just calls `Problem.copy` but I am not sure if we're worried about that code being changed (I think "No we're not" is perfectly valid).

**Review by PTNobel — APPROVED** [review]
Overall, LGTM.

I don't think we should backport semantics changes like this to old versions, so I think should be in 1.3.1 and later only.

File: cvxpy/tests/test_copy.py (hunk 2 above)
Comment (SteveDiamond) [reply cvxpy/tests/test_copy.py]: Changed to copy.copy(prob)

**Pull request merged**